contracts/src/FROSTCoordinator.sol

```diff
@@ -180,20 +197,55 @@ contract FROSTCoordinator {
         emit KeyGenSecretShared(gid, identifier, share, completed);
     }
 
-    /// @notice Submit participants secret shares. This method is the same as
-    ///         `keyGenSecretShare` with an additional callback once secret
-    ///         sharing is complete.
-    function keyGenSecretShareWithCallback(
-        FROSTGroupId.T gid,
-        KeyGenSecretShare calldata share,
-        Callback calldata callback
-    ) public returns (bool completed) {
-        completed = keyGenSecretShare(gid, share);
+    /// @notice Finalize the Key Generation ceremony for the sender if no unresolved complaints exist.
+    function keyGenFinalize(FROSTGroupId.T gid) public returns (bool completed) {
+        Group storage group = $groups[gid];
+        GroupParameters memory parameters = group.parameters;
+        require(parameters.status == GroupStatus.CONFIRMING, GroupNotCommitted());
+        FROST.Identifier identifier = group.participants.identifierOf(msg.sender);
+        group.finalizations.finalizeValidator(identifier);
+        completed = --parameters.pending == 0;
+        if (completed) {
+            parameters.status = GroupStatus.FINALIZED;
+        }
+        group.parameters = parameters;
+        emit KeyGenFinalized(gid, identifier);
+    }
+
+    /// @notice Finalize the Key Generation ceremony for the sender if no unresolved complaints exist. This method is
+    ///         the same as `keyGenFinalize` with an additional callback once finalized.
+    function keyGenFinalizeWithCallback(FROSTGroupId.T gid, Callback calldata callback)
+        public
+        returns (bool completed)
+    {
+        completed = keyGenFinalize(gid);
         if (completed) {
             callback.target.onKeyGenCompleted(gid, callback.context);
         }
     }
 
+    /// @notice Submit a complaint from the sender against another participant
+    ///         during key generation.
+    function keyGenComplain(FROSTGroupId.T gid, FROST.Identifier accused) external {
+        Group storage group = $groups[gid];
+        require(group.parameters.status == GroupStatus.CONFIRMING, GroupNotCommitted());
+        // Check if the accused identifier is valid
+        require(group.participants.getKey(accused).x != 0, InvalidAccused());
+        FROST.Identifier plaintiff = group.participants.identifierOf(msg.sender);
+        group.finalizations.submit(plaintiff, accused);
+        emit KeyGenComplained(gid, plaintiff, accused);
+    }
```

Comment:
The group needs to enter a "COMPROMISED" state in case of `threshold` complaints to a participant.